### PR TITLE
[AST] Reinstate "Fix excessive deserialization in GenericSignatureBuilder"

### PR DESCRIPTION
**Pull request by DougGregor**
Try to reinstate @slavapestov 's change to limit deserialization in `GenericSignatureBuilder`, with a fix to make `NominalTypeDecl::lookupDirect()` defensively copy its result.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

File: include/swift/AST/Decl.h [hunk 1]
```diff
@@ -2854,8 +2854,9 @@ class NominalTypeDecl : public GenericTypeDecl, public IterableDeclContext {
   ///
   /// \param ignoreNewExtensions Whether to avoid loading any new extension.
   /// Used by the module loader to break recursion.
-  ArrayRef<ValueDecl *> lookupDirect(DeclName name,
-                                     bool ignoreNewExtensions = false);
+  SmallVector<ValueDecl *, 4> lookupDirect(
```
Comment (jrose-apple) [inline include/swift/AST/Decl.h]: I think it makes sense to return a TinyPtrVector here. There are a *lot* of cases where there's just one element.

Comment (DougGregor) [inline include/swift/AST/Decl.h]: The result of this function is generally stored on the stack, so we're mostly talking saving some stack space (`TinyPtrVector` is one pointer vs. seven for the `SmallVector`). I'll make the change, though.

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Pull request merged**